mart/attack/adversary.py

```diff
@@ -19,29 +25,71 @@ class AdversaryCallbackHookMixin(Callback):
 
     callbacks = {}
 
-    def on_run_start(self, adversary, input, target, model, **kwargs):
+    def on_run_start(
+        self,
+        adversary: Callback,
```

Comment:
I think we would want to type this as `Adversary`?

Reply:
The `Adversary` class is defined below in this file, so I decided to use its parent class. Other option is to type it as `torch.nn.Module` which is the other parent class.

Reply:
Created issue #38 to tackle this. This circular issue can be a sign of a possible design problem between the adversaries and callbacks.